opensourceleg/messaging/grpc.py

```diff
@@ -0,0 +1,163 @@
+import importlib
+import math
+import os
+import time
+from concurrent import futures
+
+from opensourceleg.logging import LOGGER
+
+try:
+    import grpc
+    from google.protobuf import descriptor_pb2
+    from google.protobuf.compiler import plugin_pb2
+    from grpc_tools import protoc
+
+except ImportError:
+    LOGGER.error("grpcio-tools, types-protobuf, and grpcio must be installed.")
+
+
+class GRPCManager:
+    def __init__(self, port=50051, frequency=200):
+        self.server_data = {}
+        self.client_data = {}
+        self.callbacks = {}
+        self.proto_file = "dynamic_message.proto"
+        self.pb2_file = "dynamic_message_pb2.py"
+        self.pb2_grpc_file = "dynamic_message_pb2_grpc.py"
+        self.server = None
+        self.stub = None
+        self.port = port
+        self.pb2 = None
+        self.pb2_grpc = None
+        self.frequency = frequency
+
+    def add_variable(self, name, getter_func):
+        """Add a variable to be sent over gRPC."""
+        self.server_data[name] = getter_func
+
+    def add_callback(self, name, callback_func):
+        """Add a callback function to be called from the client."""
+        self.callbacks[name] = callback_func
+
+    def _generate_proto_file(self):
+        """Generate the proto file based on added variables."""
+        with open(self.proto_file, "w") as f:
+            f.write('syntax = "proto3";\n\n')
+            f.write("package dynamic;\n\n")
+            f.write("service DynamicService {\n")
+            f.write("  rpc SendData (Data) returns (Empty) {}\n")
+            f.write("  rpc ReceiveData (Empty) returns (stream Data) {}\n")
+            f.write("  rpc CallFunction (Function) returns (Empty) {}\n")
+            f.write("}\n\n")
+            f.write("message Data {\n")
+            f.write("  map<string, float> values = 1;\n")
+            f.write("}\n\n")
+            f.write("message Function {\n")
+            f.write("  string name = 1;\n")
+            f.write("}\n\n")
+            f.write("message Empty {}\n")
+
+    def _compile_proto_file(self):
+        """Compile the proto file to generate pb2 and pb2_grpc files."""
+        protoc.main(
+            ("", f"-I.", f"--python_out=.", f"--grpc_python_out=.", self.proto_file)
+        )
+
+    def _import_generated_modules(self):
+        """Import the dynamically generated modules."""
+        self.pb2 = importlib.import_module(self.pb2_file.split(".")[0])
+        self.pb2_grpc = importlib.import_module(self.pb2_grpc_file.split(".")[0])
+
+    def start(self):
+        """Generate proto file, compile it, and start the gRPC server."""
+        self._generate_proto_file()
+        self._compile_proto_file()
+        self._import_generated_modules()
+
+        class DynamicService(self.pb2_grpc.DynamicServiceServicer):
```

Comment:
Camel case??!

Reply:
https://protobuf.dev/programming-guides/style/

This is a style guide for .proto files that recommends CamelCase for services (in our case: SendData, ReceiveData, CallFunction), these service definitions within the .proto file then gets implemented under the DynamicService class. 


We don't really have to adhere to this style guide and do it the pythonic way with PEP-8, but if these services are being interfaced with other programming languages then it could become ugly.

Reply:
@VarunSatyadevShetty I'll leave this to you :)